torch/csrc/autograd/saved_variable.cpp

```diff
@@ -16,6 +16,24 @@ namespace torch { namespace autograd {
 
 SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_inplace_view) {
   if (variable.defined()) {
+    // Note [Inference tensor cannot be saved for backward]
+    // Invariant:
+    //   You can't save an inference tensor for backwards.
+    // If an inference tensor was saved for backward in an autograd session and
+    // then you reenter inference mode and make an inplace update to the tensor
+    // without bumping version_counter, it'll lead to silent wrong result when
+    // you do backward() for the previous autograd session. (It should error out!)
+    //
+    // Note in the documentation we say "inference tensor cannot participate
+    // in autograd" which is more restrictive than the invariant.  In practice the check is
+    // more permissive and only error out when an inference tensor is saved for
+    // backward.  Whether a tensor is saved for backward is determined by derivative
+    // formula and thus varies op by op, so by saying "no inference tensor in autograd"
+    // it's easier for users to follow.
```

Comment:
This comment is a little out of date. Technically, it's not necessary to do the test here, as given an inference tensor, we will fail when we query `current_version`. However, we do a check here anyway because we can give a better error message in this case.

The comment here also doesn't say why we don't tell users the full story in documentation. It's probably worth explicitly stating the reasoning somewhere too.